lotman/lotman_linux.go

```diff
@@ -314,9 +333,580 @@ func getFederationIssuer() (string, error) {
 	return federationIssuer, nil
 }
 
+// Given MPA1 and MPA2, merge them into a single MPA. If a field is set in MPA1,
+// it will take precedence.
+func mergeMPAs(mpa1, mpa2 *MPA) *MPA {
+	// Handle nil cases
+    if mpa1 == nil && mpa2 == nil {
+        return nil
+    }
+    if mpa1 == nil {
+        return mpa2
+    }
+    if mpa2 == nil {
+        return mpa1
+    }
+
+	// Merge individual fields
+    mergedMPA := *mpa1
+    if mpa1.DedicatedGB == nil {
+        mergedMPA.DedicatedGB = mpa2.DedicatedGB
+    }
+    if mpa1.OpportunisticGB == nil {
+        mergedMPA.OpportunisticGB = mpa2.OpportunisticGB
+    }
+    if mpa1.MaxNumObjects == nil {
+        mergedMPA.MaxNumObjects = mpa2.MaxNumObjects
+    }
+    if mpa1.CreationTime == nil {
+        mergedMPA.CreationTime = mpa2.CreationTime
+    }
+    if mpa1.ExpirationTime == nil {
+        mergedMPA.ExpirationTime = mpa2.ExpirationTime
+    }
+    if mpa1.DeletionTime == nil {
+        mergedMPA.DeletionTime = mpa2.DeletionTime
+    }
+
+    return &mergedMPA
+}
+
+// Given lot1 and lot2, merge them into a single lot. If a field is set in lot1,
+// it will take precedence. Lots cannot be merged if they have separate names
+func mergeLots(lot1, lot2 Lot) (Lot, error) {
+	if lot1.LotName != lot2.LotName {
+		return Lot{}, errors.Errorf("cannot merge lots %s and %s because they have different names", lot1.LotName, lot2.LotName)
+	}
+	mergedLot := lot1
+
+	// Prefer lot1's owner
+    if lot1.Owner == "" {
+        mergedLot.Owner = lot2.Owner
+    }
+
+	// Calculate union between the parents -- if this gets us in trouble by introducing cycles,
+	// lotman will tell us on startup (hopefully...).
+	parentSet := make(map[string]bool)
```

Comment:
I'm of the opinion that the cache administrator should get the final say in most of the configuration here. You actually made me realize that the current code prefers the discovered namespaces over the locally-configured namespaces (which I think was an oversight, corrected in one of my commits by swapping the input order of `mergeLotMaps`).

While the local admin should have final say, it'd be nice if the we still got some of the configuration from the upstream federation. Lots can only have one proper owner tied to them in the lotman database (probably a mistake at the lotman level), so I can't merge those or I think I probably would. Parents are a list so I don't have to choose between one set of config or another.

One side effect here is that the owners of a lot's parents _can_ be retrieved from lotman, which lets us treat them like proper owners. This means merging parents let us use signed tokens from parent owners to operate on lower level lots. Because we merge paths with the discovered lot namespaces, this guarantees the federation issuer still has some ability to modify lot configuration (because this setup makes every lot a child of the root lot, which is owned by the federation issuer). If this wasn't a merge, the local cache admin could effectively bar the fed administrators from managing lots, which is maybe a good thing, maybe a bad thing. Ultimately we assume a trust relationship between caches and federations, so I thought this was probably the right way to go.